src/spandrel/architectures/DAT/__init__.py

```diff
@@ -1,135 +1,90 @@
-    else:
-        upsampler = ""
-
-    num_feat_pre_layer = state_dict.get("conv_before_upsample.weight", None)
-    num_feat_layer = state_dict.get("conv_before_upsample.0.weight", None)
-    num_feat = (
-        num_feat_layer.shape[1]
-        if num_feat_layer is not None and num_feat_pre_layer is not None
-        else 64
-    )
+            # because of a head_num // 2, we can only reconstruct even head counts
+            key = f"layers.{i}.blocks.0.attn.attns.0.pos.pos3.2.weight"
+            num_heads[i] = state_dict[key].shape[0] * 2
 
-    num_in_ch = state_dict["conv_first.weight"].shape[1]
-    in_chans = num_in_ch
-    if "conv_last.weight" in state_keys:
-        num_out_ch = state_dict["conv_last.weight"].shape[0]
-    else:
-        num_out_ch = num_in_ch
-
-    upscale = 1
-    if upsampler == "nearest+conv":
-        upsample_keys = [x for x in state_keys if "conv_up" in x and "bias" not in x]
-
-        for upsample_key in upsample_keys:
-            upscale *= 2
-    elif upsampler == "pixelshuffle":
-        upsample_keys = [
-            x
-            for x in state_keys
-            if "upsample" in x and "conv" not in x and "bias" not in x
-        ]
-        for upsample_key in upsample_keys:
-            shape = state_dict[upsample_key].shape[0]
-            upscale *= math.sqrt(shape // num_feat)
-        upscale = int(upscale)
+    upsampler = (
+        "pixelshuffle" if "conv_last.weight" in state_dict else "pixelshuffledirect"
+    )
+    resi_connection = "1conv" if "conv_after_body.weight" in state_dict else "3conv"
+
+    if upsampler == "pixelshuffle":
+        upscale = 1
+        for i in range(0, get_seq_len(state_dict, "upsample"), 2):
+            num_feat = state_dict[f"upsample.{i}.weight"].shape[1]
+            shape = state_dict[f"upsample.{i}.weight"].shape[0]
+            upscale *= int(math.sqrt(shape // num_feat))
     elif upsampler == "pixelshuffledirect":
-        upscale = int(math.sqrt(state_dict["upsample.0.bias"].shape[0] // num_out_ch))
-
-    max_layer_num = 0
-    max_block_num = 0
-    for key in state_keys:
-        result = re.match(r"layers.(\d*).blocks.(\d*).norm1.weight", key)
-        if result:
-            layer_num, block_num = result.groups()
-            max_layer_num = max(max_layer_num, int(layer_num))
-            max_block_num = max(max_block_num, int(block_num))
-
-    depth = [max_block_num + 1 for _ in range(max_layer_num + 1)]
-
-    if "layers.0.blocks.1.attn.temperature" in state_keys:
-        num_heads_num = state_dict["layers.0.blocks.1.attn.temperature"].shape[0]
-        num_heads = [num_heads_num for _ in range(max_layer_num + 1)]
-    else:
-        num_heads = depth
+        num_feat = state_dict["upsample.0.weight"].shape[1]
+        upscale = int(math.sqrt(state_dict["upsample.0.weight"].shape[0] // in_chans))
 
-    embed_dim = state_dict["conv_first.weight"].shape[0]
+    qkv_bias = "layers.0.blocks.0.attn.qkv.bias" in state_dict
+
+    # old
```

Comment:
?

Reply:
I wanted to make dinner, so I just quickly finished the PR. Seems like I missed some things...